src/bigint.rs

```diff
@@ -709,77 +921,88 @@ impl Integer for BigUint {
 
-        let mut shift = 0;
-        let mut n = *other.data.last().unwrap();
-        while n < (1 << big_digit::BITS - 2) {
-            n <<= 1;
-            shift += 1;
-        }
-        assert!(shift < big_digit::BITS);
-        let (d, m) = div_mod_floor_inner(self << shift, other << shift);
-        return (d, m >> shift);
+        /*
+         * This algorithm is from Knuth, TAOCP vol 2 section 4.3, algorithm D:
+         *
+         * First, normalize the arguments so the highest bit in the highest digit of the divisor is
+         * set: the main loop uses the highest digit of the divisor for generating guesses, so we
+         * want it to be the largest number we can efficiently divide by.
+         */
 
+        let shift = other.data.last().unwrap().leading_zeros() as usize;
+        let (q, m) = div_mod_floor_inner(self << shift, other << shift);
+        return (q, m >> shift);
 
-        fn div_mod_floor_inner(a: BigUint, b: BigUint) -> (BigUint, BigUint) {
-            let mut m = a;
-            let mut d: BigUint = Zero::zero();
-            let mut n = 1;
-            while m >= b {
-                let (d0, d_unit, b_unit) = div_estimate(&m, &b, n);
-                let mut d0 = d0;
-                let mut prod = &b * &d0;
-                while prod > m {
-                    // FIXME(#5992): assignment operator overloads
-                    // d0 -= &d_unit
-                    d0   = d0 - &d_unit;
-                    // FIXME(#5992): assignment operator overloads
-                    // prod -= &b_unit;
-                    prod = prod - &b_unit
-                }
-                if d0.is_zero() {
-                    n = 2;
+
+        fn div_mod_floor_inner(mut a: BigUint, b: BigUint) -> (BigUint, BigUint) {
+            /*
+             * The algorithm works by incrementally calculating "guesses", q0, for part of the
+             * remainder. Once we have any number q0 such that q0 * b <= a, we can set
+             *
+             *     q += q0
+             *     a -= q0 * b
+             *
+             * and then iterate until a < b. Then, (q, a) will be our desired quotient and
+             * remainder.
+             *
+             * q0, our guess, is calculated by dividing the last few digits of a by the last digit
+             * of b - this should give us a guess that is "close" to the actual quotient, but is
+             * possibly greater than the actual quotient. If q0 * b > a, we simply use iterated
+             * subtraction until we have a guess such that q8 & b <= a.
+             */
+
+            let bn = *b.data.last().unwrap();
+            let q_len = a.data.len() - b.data.len() + 1;
+            let mut q: BigUint = BigUint { data: Vec::with_capacity(q_len) };
+
+            q.data.extend(repeat(0).take(q_len));
+
+            /* We reuse the same temporary to avoid hitting the allocator in our inner loop: */
+            let mut tmp: BigUint = BigUint { data: Vec::with_capacity(2) };
```

Comment:
On Mon, Nov 30, 2015 at 03:42:43PM -0800, Josh Stone wrote:
> > +             * remainder.
> > +             *
> > +             * q0, our guess, is calculated by dividing the last few digits of a by the last digit
> > +             * of b - this should give us a guess that is "close" to the actual quotient, but is
> > +             * possibly greater than the actual quotient. If q0 * b > a, we simply use iterated
> > +             * subtraction until we have a guess such that q8 & b <= a.
> > +             */
> > +
> > +            let bn = *b.data.last().unwrap();
> > +            let q_len = a.data.len() - b.data.len() + 1;
> > +            let mut q: BigUint = BigUint { data: Vec::with_capacity(q_len) };
> > +
> > +            q.data.extend(repeat(0).take(q_len));
> > +
> > +            /* We reuse the same temporary to avoid hitting the allocator in our inner loop: */
> > +            let mut tmp: BigUint = BigUint { data: Vec::with_capacity(2) };
> 
> Why 2?  Is this the worst case needed, or can that be guessed?

It's the common case, not the worst case: so, we're building up the quotient one
digit at a time, high to low - and as we add a digit q0 to the quotient we
subtract q0 * b from a.

So typically a becomes one digit shorter for every iteration of the main loop;
and as the offset into a we use for calculating our guess decreases in lockstep
with our position in the quotient, there will usually be only two digits in a
past that offset - the digits that get used for a0.

why 2 digits, and not 1? Well, working backwards - we normalized b so that bn
(the number we divide a0 by to generate our guess) was as big as it could be and
fit in a digit - so if a0 was a single digit, it'll usually be smaller than bn
and a0 / bn = (0, a0) - our guess for the quotient will be 0 - and when a digit
in the quotient is 0, a doesn't decrease in that iteration of the loop, and a0
will be one digit longer on the next iteration.

(question for myself: the algorithm assumes/requires that q0 - the guess for a
particular digit of the quotient - is not smaller than the actual quotient. but
when generating that guess, we used a0 (which is < than a) and bn (which is <
than b) - so why is q0 >= q at that digit? brb, checking knuth)